File: app/src/main/java/pl/gocards/ui/cards/list/drag_swipe/DragSwipeCardTouchHelper.java
package pl.gocards.ui.cards.list.drag_swipe;

import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Objects;

import pl.gocards.room.entity.deck.Card;
import pl.gocards.ui.cards.list.standard.CardViewHolder;

/**
 * C_U_03 Dragging the card to another position.
 * C_D_04 Delete the card by swiping.
 *
 * @author Grzegorz Ziemski
 */
public class DragSwipeCardTouchHelper extends ItemTouchHelper.Callback {

    protected static final int NO_DRAG = -1;
    protected static final int DRAG_AFTER_POSITION = 1;

    private final DragSwipeListCardsAdapter adapter;
    private int dragToPosition = NO_DRAG;

    public DragSwipeCardTouchHelper(DragSwipeListCardsAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public void clearView(
            @NonNull RecyclerView recyclerView,
            @NonNull RecyclerView.ViewHolder viewHolder
    ) {
        super.clearView(recyclerView, viewHolder);
        CardViewHolder cardViewHolder = (CardViewHolder) viewHolder;
        cardViewHolder.unfocusItemView();
        dragToPosition = NO_DRAG;
    }

    @Override
    public void onSelectedChanged(
            @Nullable RecyclerView.ViewHolder viewHolder,
            int actionState
    ) {
        super.onSelectedChanged(viewHolder, actionState);
        CardViewHolder cardViewHolder = (CardViewHolder) viewHolder;
        switch (actionState) {
            case ItemTouchHelper.ACTION_STATE_DRAG, ItemTouchHelper.ACTION_STATE_SWIPE ->
                    Objects.requireNonNull(cardViewHolder).focusSingleTapItemView();
            case ItemTouchHelper.ACTION_STATE_IDLE -> {
                if (dragToPosition != -1) {
                    Card card = adapter.getItem(dragToPosition);
                    if (card.getOrdinal() != dragToPosition + DRAG_AFTER_POSITION) {
                        adapter.onMovedCard(card, dragToPosition + DRAG_AFTER_POSITION);
                    }
                }
            }
        }
    }

    @Override
    public boolean onMove(
            @NonNull RecyclerView recyclerView,
            @NonNull RecyclerView.ViewHolder viewHolder,
            @NonNull RecyclerView.ViewHolder target
    ) {
        dragToPosition = target.getBindingAdapterPosition();
        adapter.onMoveCard(viewHolder.getBindingAdapterPosition(), dragToPosition);
        return true;
    }

    /**
     * C_D_04 Delete the card by swiping.
     */
    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
        adapter.onClickDeleteCard(viewHolder.getBindingAdapterPosition());
    }

    /**
     * It is implemented by {@link DragSwipeCardViewHolder#onLongPress(MotionEvent)}
     */
    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public int getMovementFlags(
            @NonNull RecyclerView recyclerView,
            @NonNull RecyclerView.ViewHolder viewHolder
    ) {
        final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        final int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
        return makeMovementFlags(dragFlags, swipeFlags);
    }

    protected DragSwipeListCardsAdapter getAdapter() {
        return adapter;
    }
}

File: app/src/main/java/pl/gocards/ui/cards/kt/list/drag_swap/DragSwipeCardTouchHelper.kt
package pl.gocards.ui.cards.kt.list.drag_swap

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import pl.gocards.ui.cards.kt.list.display.CardViewHolder

/**
 * C_U_03 Dragging the card to another position.
 * C_D_04 Delete the card by swiping.
 *
 * @author Grzegorz Ziemski
 */
open class DragSwipeCardTouchHelper(
    open val adapter: DragSwipeListCardsAdapter
): ItemTouchHelper.Callback() {

    companion object {
        private const val NO_DRAG = -1
    }

    private var dragToPosition = NO_DRAG

    override fun clearView(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ) {
        super.clearView(recyclerView, viewHolder)
        val position = viewHolder.bindingAdapterPosition
        if (position > -1) {
            adapter.setActive(viewHolder.bindingAdapterPosition, false)
            (viewHolder as CardViewHolder).unfocus()
            dragToPosition = NO_DRAG
        }
    }

    override fun onSelectedChanged(
        viewHolder: RecyclerView.ViewHolder?,
        actionState: Int
    ) {
        super.onSelectedChanged(viewHolder, actionState)

        when (actionState) {
            ItemTouchHelper.ACTION_STATE_DRAG,
            ItemTouchHelper.ACTION_STATE_SWIPE -> {
                val index = viewHolder?.bindingAdapterPosition ?: return
                adapter.setActive(index, true)
                (viewHolder as CardViewHolder).focus()
            }
            ItemTouchHelper.ACTION_STATE_IDLE -> {
                if (dragToPosition != -1) {
                    adapter.moveDb(dragToPosition, dragToPosition + 1)
                }
            }
        }
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean {
        dragToPosition = target.bindingAdapterPosition
        adapter.move(viewHolder.bindingAdapterPosition, dragToPosition)
        return true
    }

    /**
     * C_D_04 Delete the card by swiping.
     */
    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        adapter.delete(viewHolder.bindingAdapterPosition)
    }

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ): Int {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
        val swipeFlags = ItemTouchHelper.START or ItemTouchHelper.END
        return makeMovementFlags(dragFlags, swipeFlags)
    }
}